lib/util/content_steering_manager.js

```diff
@@ -0,0 +1,364 @@
+      queryData.add('_HLS_throughput', String(currentBandwidth));
+    }
+    if (queryData.getCount()) {
+      finalUri.setQueryData(queryData);
+    }
+    return finalUri.toString();
+  }
+
+  /**
+   * @param {shaka.util.ContentSteeringManager.SteeringManifest} manifest
+   * @param {string} finalManifestUri
+   * @private
+   */
+  processManifest_(manifest, finalManifestUri) {
+    if (this.updateTimer_ != null) {
+      this.updateTimer_.stop();
+      this.updateTimer_ = null;
+    }
+    const uri = manifest['RELOAD-URI'] || finalManifestUri;
+    this.updateTimer_ = new shaka.util.Timer(() => {
+      this.requestInfo(uri);
+    });
+    if (manifest.TTL) {
+      this.lastTTL_ = 1;
+    }
+    this.updateTimer_.tickAfter(this.lastTTL_);
+    this.pathwayPriority_ = manifest['PATHWAY-PRIORITY'] || [];
+    this.pathwayClones_ = manifest['PATHWAY-CLONES'] || [];
+  }
+
+  /**
+   * Clear the previous locations added.
+   */
+  clearPreviousLocations() {
+    this.locations_.clear();
+  }
+
+  /**
+   * @param {string} streamId
+   * @param {string} pathwayId
+   * @param {string} uri
+   */
+  addLocation(streamId, pathwayId, uri) {
+    let streamLocations = this.locations_.get(streamId);
+    if (!streamLocations) {
+      streamLocations = new Map();
+    }
+    streamLocations.set(pathwayId, uri);
+    this.locations_.set(streamId, streamLocations);
+  }
+
+  /**
+   * @param {string} uri
+   */
+  banLocation(uri) {
+    const bannedUntil = Date.now() + 60000;
+    this.bannedLocations_.set(uri, bannedUntil);
+  }
+
+  /**
+   * Get the base locations ordered according the priority.
+   *
+   * @param {string} streamId
+   * @return {!Array.<string>}
+   */
+  getLocations(streamId) {
+    const streamLocations = this.locations_.get(streamId) || new Map();
+    let locationsMap = [];
+    for (const pathwayId of this.pathwayPriority_) {
+      const location = streamLocations.get(pathwayId);
+      if (location) {
+        locationsMap.push({
+          pathwayId: pathwayId,
+          location: location,
+        });
+      } else {
+        const clone = this.pathwayClones_.find((c) => c.ID == pathwayId);
+        if (clone) {
+          const cloneLocation = streamLocations.get(clone['BASE-ID']);
+          if (cloneLocation) {
+            if (clone['URI-REPLACEMENT'].HOST) {
+              const uri = new goog.Uri(cloneLocation);
+              uri.setDomain(clone['URI-REPLACEMENT'].HOST);
+              locationsMap.push({
+                pathwayId: pathwayId,
+                location: uri.toString(),
+              });
+            } else {
+              locationsMap.push({
+                pathwayId: pathwayId,
+                location: cloneLocation,
+              });
+            }
+          }
+        }
+      }
+    }
+
+    if (locationsMap.length && this.bannedLocations_.size) {
+      const now = Date.now();
+      const toDelete = [];
+      for (const [uri, bannedUntil] of this.bannedLocations_) {
```

Comment:
Couldn't you do something like...
```js
for (const uri of this.bannedLocations_.keys()) {
  const bannedUntil = this.bannedLocations_.get(uri);
  if (now > bannedUntil) {
    this.bannedLocations_.delete(uri);
  }
}
locationsMap = locationsMap.filter((l) => {
  for (const uri of this.bannedLocations_.keys()) {
    if (uri.includes(l.location)) {
      return false;
    }
  }
  return true;
});
```
I think that would be both shorter, and more readable since it breaks the "removing old banned locations" and "applying banned locations" parts into separate clauses.

Reply:
Changed!